src/passes/HeapStoreOptimization.cpp

```diff
@@ -246,6 +283,84 @@ struct HeapStoreOptimization
       }
     }
 
+    // We must also be careful of branches out from the value. For example:
+    //
+    //  (block $out
+    //    (local.set $x (struct.new X Y Z))
+    //    (struct.set (local.get $x) (..br $out..))  ;; X' here has a br
+    //  )
+    //  ..use $x..
+    //
+    // Note how we do the local.set first. Imagine we optimized to this:
+    //
+    //  (block $out
+    //    (local.set $x (struct.new (..br $out..) Y Z))
+    //  )
+    //  ..use $x..
+    //
+    // Now the br happens first, skipping the local.set entirely, and the use
+    // later down will not get the proper value.
+    //
+    // To check for this problem, see which basic blocks can be reached from the
+    // value. The structured IR limits what is possible here, since we know that
+    // the struct.set follows the local.set: the only problem is one like we
+    // showed in the example, where we branch to something after the
+    // struct.set's basic block, in post-order. As we are processing the blocks
```

Comment:
Why does the direction of the branch matter? Wouldn't we have a problem if `$x` is live at the branch target, no matter where the branch is going? It would be good to add a test that uses a loop.

Reply:
Good catch, yeah, this does need more work to handle loops. I'll look into it.